_posts/2022-07-06-#25077.md

```diff
@@ -46,9 +46,168 @@ commit:
+17:26 <MacroFake> afmencken: Correct
+17:27 <MacroFake> (Still q 3), next step: Why is cs_main not held?
+17:28 <MacroFake> Hint: The implemenation of ActiveTip: https://github.com/bitcoin/bitcoin/blob/aeab1b42e67cc8146bfc7d127d15633bd652fe60/src/validation.h#L914-L918
+17:29 <MacroFake> And the implementation of ActiveChainState: https://github.com/bitcoin/bitcoin/blob/aeab1b42e67cc8146bfc7d127d15633bd652fe60/src/validation.cpp#L5157
+17:29 <michaelfolkson> Why was cs_main not held before this PR? Is that the question, sorry
+17:29 <larryruane> "Why is cs_main not held?" -- is it anything more than that it's an oversight? Ah I think I see, those first 3 methods you linked to don't scquire the lock
+17:29 <glozow> cs_main is released at the end of `ActiveChainState()` ?
+17:30 <MacroFake> michaelfolkson: Yes, that is the question
+17:31 <michaelfolkson> Because it is held by a different thread?
+17:31 <larryruane> (like `ActiveChainState` does... they should be similar in their locking behavior)
+17:31 <MacroFake> glozow: Correct. The cs_main Mutex will only be held until the scope is closed. The scope in this case is the function body {...}
+17:32 <MacroFake> michaelfolkson: If a mutex is acquired by a different thread, it can not be acquired by this thread. However, if the thread *wants* to acquire it, it will wait until it can
+17:32 <lightlike> is there still a "race", just not one with UB? After the fix, loadblk still does its thing and possible changes the ActiveTip() several times, and at some point the main thread takes the cs_main and calls RPCNotifyBlockChange() with whatever is the current tip at this point - could that vary from run to run?
+17:32 <MacroFake> michaelfolkson: If the thread does not want to acquire it, it will simply continue
+17:33 <larryruane> ignore if this is too detailed, but "... it will wait until it can" -- does it spin-wait, or sleep-wait? Or spin for a short time and then "give up" and sleep (let other threads run)?
+17:33 <MacroFake> lightlike: Good point. I think not, because RPCNotifyBlockChange doesn't care about the exact block, but I should double-check.
+17:35 <MacroFake> larryruane: I am not familiar with low level primitives, but I'd guess that the thread will be put to sleep and woken when the lock is ready to be picked up
+17:35 <sipa> larryruane: In Linux these days, locks are implemented using a mechanism called "futexes" (which have their own wikipedia article)
+17:35 <larryruane> MacroFake: sipa: thanks
+17:35 <MacroFake> (final part of q 3): How is the data race fixed in the pull request? 
+17:36 <MacroFake> sipa: Thanks!
+17:36 <sipa> IIRC the basic idea is to use atomic CPU instructions to quickly check if the lock is contended or not, and if it is, make a kernel call that gets woken up when needed
+17:36 <larryruane> fixed by: https://github.com/bitcoin/bitcoin/pull/25077/files#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1856
```

Comment:
This link is now broken, but I don't remember which line it was (something where the diff added a `WITH_LOCK()`). If @LarryRuane remembers maybe you could point me to which line of code it is in [this branch](https://github.com/bitcoin-core-review-club/bitcoin/tree/pr25077)? No worries if you don't remember.

For now, I will merge as is. This is an exercise to the log reader.